src/services/item/item.repository.ts

```diff
@@ -277,6 +279,82 @@ export class ItemRepository {
     }));
   }
 
+  async getParentsForPublic(dbConnection: DBConnection, item: ItemRaw) {
+    const itemTree = await dbConnection
+      .select()
+      .from(items)
+      .where(isAncestorOrSelf(items.path, item.path))
+      .as('item_tree');
+
+    const ivTree = await dbConnection
+      .select()
+      .from(itemVisibilitiesTable)
+      .where(isAncestorOrSelf(itemVisibilitiesTable.itemPath, item.path))
+      .as('iv_tree');
+
+    const conditions = and(ne(itemTree.id, item.id), or(eq(ivTree.type, 'public')));
+
+    const result = dbConnection
+      .select()
+      .from(itemTree)
+      .leftJoin(ivTree, isAncestorOrSelf(ivTree.itemPath, itemTree.path))
+      .where(conditions)
+      .orderBy(asc(sql`nlevel(path)`));
+
+    return (await result).map(({ item_tree }) => ({
+      ...item_tree,
+    }));
+  }
+
+  async getParentsForAccount(dbConnection: DBConnection, item: ItemRaw, user: AuthenticatedUser) {
+    const itemTree = await dbConnection
+      .select()
+      .from(items)
+      .where(isAncestorOrSelf(items.path, item.path))
+      .as('item_tree');
+
+    const imTree = await dbConnection
+      .select()
+      .from(itemMembershipsTable)
+      .where(
+        and(
+          isAncestorOrSelf(itemMembershipsTable.itemPath, item.path),
+          eq(itemMembershipsTable.accountId, user.id),
+        ),
+      )
+      .as('im_tree');
+
+    const ivTree = await dbConnection
+      .select()
+      .from(itemVisibilitiesTable)
+      .where(isAncestorOrSelf(itemVisibilitiesTable.itemPath, item.path))
+      .as('iv_tree');
+
+    const conditions = and(
+      ne(itemTree.id, item.id),
+      or(
+        eq(imTree.permission, 'admin'),
+        eq(imTree.permission, 'write'),
+        eq(ivTree.type, 'public'),
+        and(ne(ivTree.type, 'hidden'), eq(imTree.permission, 'read')),
+        and(isNull(ivTree.type), eq(imTree.permission, 'read')),
+      ),
+    );
+
+    const result = dbConnection
+      .select()
+      .from(itemTree)
+      .leftJoin(imTree, isAncestorOrSelf(imTree.itemPath, itemTree.path))
+      .leftJoin(ivTree, isAncestorOrSelf(ivTree.itemPath, itemTree.path))
+      .where(conditions)
+      .orderBy(asc(sql`nlevel(path)`));
+
+    return (await result).map(({ item_tree }) => ({
```

Comment:
I am not a big fan of doing the await in parenthesis like this. I think for legibility it would be better to directly await the result. Maybe add a comment that tells why we await here and not in other queries in this function.